File: src/test/java/com/lalkalol/mujahid/commands/impl/PlaylistCommandTest.java
package com.lalkalol.mujahid.commands.impl;

import com.lalkalol.mujahid.commands.CommandTestFixtures;
import com.lalkalol.mujahid.db.PlaylistSummary;
import com.lalkalol.mujahid.db.StoredTrack;
import dev.arbjerg.lavalink.client.player.Track;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.ArgumentMatchers.anyList;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

class PlaylistCommandTest {

    @Test
    void save_rejectsEmptyName() {
        var h = CommandTestFixtures.harness()
                .withStringOption("name", "   ");
        when(h.event.getSubcommandName()).thenReturn("save");

        new PlaylistCommand().execute(h.context());

        assertTrue(h.lastEphemeralDescription().contains("cannot be empty"));
    }

    @Test
    void save_rejectsWhenNothingToSave() {
        var h = CommandTestFixtures.harness()
                .withStringOption("name", "favorites");
        when(h.event.getSubcommandName()).thenReturn("save");
        when(h.scheduler.getCurrent()).thenReturn(null);
        when(h.scheduler.queueSnapshot()).thenReturn(List.of());

        new PlaylistCommand().execute(h.context());

        assertTrue(h.lastEphemeralDescription().contains("nothing playing or queued"));
    }

    @Test
    void save_persistsCurrentTrackAndQueue() {
        var h = CommandTestFixtures.harness()
                .withStringOption("name", "favorites");
        when(h.event.getSubcommandName()).thenReturn("save");

        Track current = h.mockTrack("Now");
        Track queued = h.mockTrack("Later");
        when(h.scheduler.getCurrent()).thenReturn(current);
        when(h.scheduler.queueSnapshot()).thenReturn(List.of(queued));
        when(h.playlists.save(eq(CommandTestFixtures.USER_ID), eq("favorites"), anyList())).thenReturn(2);

        new PlaylistCommand().execute(h.context());

        verify(h.playlists).save(eq(CommandTestFixtures.USER_ID), eq("favorites"), anyList());
        assertTrue(h.lastReplyDescription().contains("favorites"));
        assertTrue(h.lastReplyDescription().contains("2"));
    }

    @Test
    void load_rejectsMissingPlaylist() {
        var h = CommandTestFixtures.harness()
                .withStringOption("name", "missing");
        when(h.event.getSubcommandName()).thenReturn("load");
        when(h.playlists.load(CommandTestFixtures.USER_ID, "missing")).thenReturn(null);

        new PlaylistCommand().execute(h.context());

        assertTrue(h.lastEphemeralDescription().contains("missing"));
    }

    @Test
    void list_showsSavedPlaylists() {
        var h = CommandTestFixtures.harness();
        when(h.event.getSubcommandName()).thenReturn("list");
        when(h.playlists.list(CommandTestFixtures.USER_ID)).thenReturn(
                List.of(new PlaylistSummary("chill", 4), new PlaylistSummary("party", 10))
        );

        new PlaylistCommand().execute(h.context());

        assertTrue(h.lastEphemeralDescription().contains("chill"));
        assertTrue(h.lastEphemeralDescription().contains("party"));
    }

    @Test
    void list_rejectsWhenNoPlaylists() {
        var h = CommandTestFixtures.harness();
        when(h.event.getSubcommandName()).thenReturn("list");
        when(h.playlists.list(CommandTestFixtures.USER_ID)).thenReturn(List.of());

        new PlaylistCommand().execute(h.context());

        assertTrue(h.lastEphemeralDescription().contains("no saved playlists"));
    }

    @Test
    void delete_removesExistingPlaylist() {
        var h = CommandTestFixtures.harness()
                .withStringOption("name", "old");
        when(h.event.getSubcommandName()).thenReturn("delete");
        when(h.playlists.delete(CommandTestFixtures.USER_ID, "old")).thenReturn(true);

        new PlaylistCommand().execute(h.context());

        assertTrue(h.lastReplyDescription().contains("Deleted"));
    }

    @Test
    void delete_reportsMissingPlaylist() {
        var h = CommandTestFixtures.harness()
                .withStringOption("name", "ghost");
        when(h.event.getSubcommandName()).thenReturn("delete");
        when(h.playlists.delete(CommandTestFixtures.USER_ID, "ghost")).thenReturn(false);

        new PlaylistCommand().execute(h.context());

        assertTrue(h.lastEphemeralDescription().contains("ghost"));
    }

    @Test
    void load_rejectsEmptyStoredPlaylist() {
        var h = CommandTestFixtures.harness()
                .withStringOption("name", "empty");
        when(h.event.getSubcommandName()).thenReturn("load");
        when(h.playlists.load(CommandTestFixtures.USER_ID, "empty")).thenReturn(List.of());

        new PlaylistCommand().execute(h.context());

        assertTrue(h.lastEphemeralDescription().contains("empty"));
    }

    @Test
    void save_mapsTracksToStoredFormat() {
        var h = CommandTestFixtures.harness()
                .withStringOption("name", "mix");
        when(h.event.getSubcommandName()).thenReturn("save");
        var current = h.mockTrack("A");
        when(h.scheduler.getCurrent()).thenReturn(current);
        when(h.scheduler.queueSnapshot()).thenReturn(List.of());
        when(h.playlists.save(eq(CommandTestFixtures.USER_ID), eq("mix"), anyList())).thenAnswer(inv -> {
            @SuppressWarnings("unchecked")
            List<StoredTrack> tracks = inv.getArgument(2);
            return tracks.size();
        });

        new PlaylistCommand().execute(h.context());

        assertTrue(h.lastReplyDescription().contains("1"));
    }
}

File: src/test/kotlin/com/lalkalol/mujahid/commands/impl/PlaylistCommandTest.kt
package com.lalkalol.mujahid.commands.impl

import com.lalkalol.mujahid.commands.CommandTestFixtures
import com.lalkalol.mujahid.db.PlaylistSummary
import com.lalkalol.mujahid.db.StoredTrack
import kotlinx.coroutines.runBlocking
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.mockito.kotlin.any
import org.mockito.kotlin.eq
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever

class PlaylistCommandTest {
    @Test
    fun save_rejectsEmptyName() = runBlocking {
        val h = CommandTestFixtures.harness()
            .withStringOption("name", "   ")
        whenever(h.event.subcommandName).thenReturn("save")

        PlaylistCommand().execute(h.context())

        assertTrue(h.lastEphemeralDescription().contains("cannot be empty"))
    }

    @Test
    fun save_rejectsWhenNothingToSave() = runBlocking {
        val h = CommandTestFixtures.harness()
            .withStringOption("name", "favorites")
        whenever(h.event.subcommandName).thenReturn("save")
        whenever(h.scheduler.getCurrent()).thenReturn(null)
        whenever(h.scheduler.queueSnapshot()).thenReturn(emptyList())

        PlaylistCommand().execute(h.context())

        assertTrue(h.lastEphemeralDescription().contains("nothing playing or queued"))
    }

    @Test
    fun save_persistsCurrentTrackAndQueue() = runBlocking {
        val h = CommandTestFixtures.harness()
            .withStringOption("name", "favorites")
        whenever(h.event.subcommandName).thenReturn("save")

        val current = h.mockTrack("Now")
        val queued = h.mockTrack("Later")
        whenever(h.scheduler.getCurrent()).thenReturn(current)
        whenever(h.scheduler.queueSnapshot()).thenReturn(listOf(queued))
        whenever(h.playlists.save(eq(CommandTestFixtures.USER_ID), eq("favorites"), any())).thenReturn(2)

        PlaylistCommand().execute(h.context())

        verify(h.playlists).save(eq(CommandTestFixtures.USER_ID), eq("favorites"), any())
        assertTrue(h.lastReplyDescription().contains("favorites"))
        assertTrue(h.lastReplyDescription().contains("2"))
    }

    @Test
    fun load_rejectsMissingPlaylist() = runBlocking {
        val h = CommandTestFixtures.harness()
            .withStringOption("name", "missing")
        whenever(h.event.subcommandName).thenReturn("load")
        whenever(h.playlists.load(CommandTestFixtures.USER_ID, "missing")).thenReturn(null)

        PlaylistCommand().execute(h.context())

        assertTrue(h.lastEphemeralDescription().contains("missing"))
    }

    @Test
    fun list_showsSavedPlaylists() = runBlocking {
        val h = CommandTestFixtures.harness()
        whenever(h.event.subcommandName).thenReturn("list")
        whenever(h.playlists.list(CommandTestFixtures.USER_ID)).thenReturn(
            listOf(PlaylistSummary("chill", 4), PlaylistSummary("party", 10)),
        )

        PlaylistCommand().execute(h.context())

        assertTrue(h.lastEphemeralDescription().contains("chill"))
        assertTrue(h.lastEphemeralDescription().contains("party"))
    }

    @Test
    fun list_rejectsWhenNoPlaylists() = runBlocking {
        val h = CommandTestFixtures.harness()
        whenever(h.event.subcommandName).thenReturn("list")
        whenever(h.playlists.list(CommandTestFixtures.USER_ID)).thenReturn(emptyList())

        PlaylistCommand().execute(h.context())

        assertTrue(h.lastEphemeralDescription().contains("no saved playlists"))
    }

    @Test
    fun delete_removesExistingPlaylist() = runBlocking {
        val h = CommandTestFixtures.harness()
            .withStringOption("name", "old")
        whenever(h.event.subcommandName).thenReturn("delete")
        whenever(h.playlists.delete(CommandTestFixtures.USER_ID, "old")).thenReturn(true)

        PlaylistCommand().execute(h.context())

        assertTrue(h.lastReplyDescription().contains("Deleted"))
    }

    @Test
    fun delete_reportsMissingPlaylist() = runBlocking {
        val h = CommandTestFixtures.harness()
            .withStringOption("name", "ghost")
        whenever(h.event.subcommandName).thenReturn("delete")
        whenever(h.playlists.delete(CommandTestFixtures.USER_ID, "ghost")).thenReturn(false)

        PlaylistCommand().execute(h.context())

        assertTrue(h.lastEphemeralDescription().contains("ghost"))
    }

    @Test
    fun load_rejectsEmptyStoredPlaylist() = runBlocking {
        val h = CommandTestFixtures.harness()
            .withStringOption("name", "empty")
        whenever(h.event.subcommandName).thenReturn("load")
        whenever(h.playlists.load(CommandTestFixtures.USER_ID, "empty")).thenReturn(emptyList())

        PlaylistCommand().execute(h.context())

        assertTrue(h.lastEphemeralDescription().contains("empty"))
    }

    @Test
    fun save_mapsTracksToStoredFormat() = runBlocking {
        val h = CommandTestFixtures.harness()
            .withStringOption("name", "mix")
        whenever(h.event.subcommandName).thenReturn("save")
        val current = h.mockTrack("A")
        whenever(h.scheduler.getCurrent()).thenReturn(current)
        whenever(h.scheduler.queueSnapshot()).thenReturn(emptyList())
        whenever(h.playlists.save(eq(CommandTestFixtures.USER_ID), eq("mix"), any())).thenAnswer { invocation ->
            @Suppress("UNCHECKED_CAST")
            val tracks = invocation.getArgument<List<StoredTrack>>(2)
            tracks.size
        }

        PlaylistCommand().execute(h.context())

        assertTrue(h.lastReplyDescription().contains("1"))
    }
}
